R/busy-indicators.R

```diff
@@ -0,0 +1,229 @@
+#' this function in the app's UI.
+#'
+#' @param ... Currently ignored (for future expansion).
+#' @param spinner_type The type of spinner. Pre-bundled types include:
+#'   '`r paste0(.busySpinnerTypes, collapse = "', '")`'.
+#'
+#'   A path to a local SVG file can also be provided. The SVG should adhere to
+#'   the following rules:
+#'   * The SVG itself should contain the animation.
+#'   * It should avoid absolute sizes (the spinner's containing DOM element
+#'     size is set in CSS by `spinner_size`, so it should fill that container).
+#'   * It should avoid setting absolute colors (the spinner's containing DOM element
+#'     color is set in CSS by `spinner_color`, so it should inherit that color).
+#' @param spinner_color The color of the spinner. This can be any valid CSS
+#'   color. Defaults to the app's "primary" color if Bootstrap is on the page.
+#' @param spinner_size The size of the spinner. This can be any valid CSS size.
+#' @param spinner_delay The amount of time to wait before showing the spinner.
+#'   This can be any valid CSS time and can be useful for not showing the spinner
+#'   if the computation finishes quickly.
+#' @param spinner_selector A character string containing a CSS selector for
+#'   scoping the spinner customization. The default (`NULL`) applies the
+#'   customization globally (i.e., to the root DOM element). Can also be set to
+#'   `NA` to apply the customization as inline styles to a specific element
+#'   (e.g., `div(busyIndicatorOptions(..., spinner_selector = NA))`).
+#' @param pulse_background A CSS background definition for the pulse. The
+#'   default uses a
+#'   [linear-gradient](https://developer.mozilla.org/en-US/docs/Web/CSS/gradient/linear-gradient)
+#'   of the theme's indigo, purple, and pink colors.
+#' @param pulse_height The height of the pulsing banner. This can be any valid
+#'   CSS size.
+#' @param pulse_speed The speed of the pulsing banner. This can be any valid CSS
+#'   time.
+#'
+#' @export
+#' @seealso [useBusyIndicators()] for enabling/disabling busy indicators.
+#' @examplesIf rlang::is_interactive()
+#'
+#' library(bslib)
+#'
+#' ui <- page_fillable(
+#'   useBusyIndicators(),
+#'   busyIndicatorOptions(
+#'     spinner_type = "bars",
+#'     spinner_color = "indigo",
+#'     spinner_size = "4rem"
+#'   ),
+#'   card(
+#'     card_header(
+#'       "A plot",
+#'       input_task_button("simulate", "Simulate"),
+#'       class = "d-flex justify-content-between align-items-center"
+#'     ),
+#'     plotOutput("p"),
+#'   )
+#' )
+#'
+#' server <- function(input, output) {
+#'   output$p <- renderPlot({
+#'     input$simulate
+#'     Sys.sleep(4)
+#'     plot(x = rnorm(100), y = rnorm(100))
+#'   })
+#' }
+#'
+#' shinyApp(ui, server)
+busyIndicatorOptions <- function(
+  ...,
+  spinner_type = NULL,
+  spinner_color = NULL,
+  spinner_size = NULL,
+  spinner_delay = NULL,
+  spinner_selector = NULL,
```

Comment:
Two thoughts here:

1. I think we should use `spinner_selector = ":root"` as the default to communicate the default value. I prefer to put the default value in the function signature unless there's a strong reason why we'd hide it (the default is calculated or complicated).

2. If `spinner_selector` has a default value, it seems a whole lot easier to use `NULL` as the signal instead of `NA`. I'd argue it's more consistent: then in every argument `NULL` is "no selection", where for `spinner_selector` "no selection" means no selector which means "give me inline styles".
  
    As it is now, `NULL` means "no selection" for every option _except_ `spinner_selector`, where it currently means "use the default".

    That said, I can live with `NA`, but I think `spinnerOptions()` should require that `spinner_selector` is a scalar string if not NA.

Reply:
> I'd argue it's more consistent

In that sense, yes, but it's also less consistent in that an input value of `NULL` for that particular argument gives you a fundamentally different return value.

Reply:
Agreed, but that inconsistency exists regardless of which sentinel value we use for `spinner_selector`.

Reply:
I can live with `NA` if you feel strongly about it, but I _do feel strongly_ that we shouldn't use `NULL` as the default. If the default value is `:root` we should just tell people that and not hide it from them.